sdk/core/azure-core/azure/core/pipeline/policies/_base_async.py

```diff
@@ -60,7 +63,7 @@ class AsyncHTTPPolicy(abc.ABC, Generic[HTTPRequestType, AsyncHTTPResponseType]):
     """
     def __init__(self) -> None:
         # next will be set once in the pipeline
-        self.next = None # type: Optional[Any]
+        self.next = None # type: Optional[Union[AsyncHTTPPolicy, AsyncHttpTransport]]
```

Comment:
In actual usage AsyncPipeline ensures `self.next` is not None, so I think it's reasonable for this not to be Optional. (It matters because subclasses will get mypy errors on `self.next.send()` if this is Optional.)

Reply:
But we set it to None here, right?

If I set it non-nullable, mypy complains here.

Reply:
Yes, but it won't be None when used by AsyncPipeline. Typing it as non-Optional saves subclasses from doing something clumsy purely to satisfy the type checker ([for example](https://github.com/Azure/azure-sdk-for-python/pull/15832/files#diff-e17dd413b7adbbeb7dfb061ab45d44e514003a1079c729e192636df9bea72a00R90)) that doesn't reflect real usage. Perhaps the question is, Do we expect AsyncHTTPPolicy to be used without AsyncPipeline? If yes, the type should be Optional because subclasses can't assume someone else assigned a non-None value. But I think the answer is "no", and incidentally that's evidently the answer for [HTTPPolicy](https://github.com/Azure/azure-sdk-for-python/blob/c4b9e911b78c3adc3ea3a0a43d43421d5dd7d60f/sdk/core/azure-core/azure/core/pipeline/policies/_base.py#L64).

Reply:
To be honest, I don't know why it does not complain with HTTPPolicy while we set it to None but declare it is not optional. (it seems to me not a correct behavior).

If we want to declare self.next is not optional, we have to give it a value here. Either AsyncHTTPPolicy or AsyncHttpTransport.

Given this is the AsyncHTTPPolicy class and Python does not support pointer. self.next = AsyncHTTPPolicy() will cause dead loop.

AsyncHttpTransport is an abstract class. Do you think self.next = AiohttpTransport() is a good idea?

Reply:
Seems this is something I don't understand either. There's no error here:
```py
class A:
    pass

class B:
    def __init__(self):
        self.thing = None  # type: A

reveal_type(B().thing)  # Revealed type is A
```
Assigning `self.next` an instance of something seems excessive to me as a means to override mypy's judgment. Using `cast` is ugly but its purpose is clear, it accomplishes the same thing, and has no runtime implications:
```py
self.next = cast(Union[AsyncHTTPPolicy, AsyncHttpTransport], None)
```